cmd/consensus/main.go

```diff
@@ -241,91 +242,68 @@ func main() {
-			// root block)
+			// If the node has a beacon key file, then save it to the secrets database
+			// as the beacon key for the first epoch of the root snapshot.
 			beaconPrivateKey, err = loadBeaconPrivateKey(node.BaseConfig.BootstrapDir, node.NodeID)
+			if errors.Is(err, os.ErrNotExist) {
+				return fmt.Errorf("node is starting from spork root snapshot, but does not have spork random beacon key file: %w", err)
+			}
 			if err != nil {
-				return err
+				return fmt.Errorf("could not load beacon key file: %w", err)
 			}
+
+			rootEpoch := node.State.AtBlockID(node.RootBlock.ID()).Epochs().Current()
 			epochCounter, err := rootEpoch.Counter()
 			if err != nil {
-				return err
-			}
-			err = dkgState.InsertMyBeaconPrivateKey(epochCounter, beaconPrivateKey.PrivateKey)
-			if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
-				return err
-			}
-			// mark the root DKG as successful, so it is considered safe
-			err = dkgState.SetDKGEndState(epochCounter, flow.DKGEndStateSuccess)
-			if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
-				return err
+				return fmt.Errorf("could not get root epoch counter: %w", err)
 			}
 
-			// Given an epoch, checkEpochKey returns an error if we are a
-			// participant in the epoch and we don't have the corresponding DKG
-			// key in the database.
-			checkEpochKey := func(protocol.Epoch) error {
-				identities, err := rootEpoch.InitialIdentities()
-				if err != nil {
-					return err
-				}
-				if _, ok := identities.ByNodeID(node.NodeID); ok {
-					counter, err := rootEpoch.Counter()
-					if err != nil {
-						return err
-					}
-					_, err = dkgState.RetrieveMyBeaconPrivateKey(counter)
-					if err != nil {
-						return err
-					}
-				}
-				return nil
+			// confirm the beacon key file matches the canonical public keys
+			rootDKG, err := rootEpoch.DKG()
+			if err != nil {
+				return fmt.Errorf("could not get dkg for root epoch: %w", err)
 			}
-
-			// if we are a member of the current epoch, make sure we have the
-			// DKG key
-			currentEpoch := node.State.Final().Epochs().Current()
-			err = checkEpochKey(currentEpoch)
+			myBeaconPublicKeyShare, err := rootDKG.KeyShare(node.NodeID)
 			if err != nil {
-				return fmt.Errorf("a random beacon that we are a participant in is currently in use and we don't have our key share for it: %w", err)
+				return fmt.Errorf("could not get my beacon public key share for root epoch: %w", err)
 			}
 
-			// if we participated in the DKG protocol for the next epoch, and we
-			// are in EpochCommitted phase, make sure we have saved the
-			// resulting DKG key
-			phase, err := node.State.Final().Phase()
-			if err != nil {
+			if !myBeaconPublicKeyShare.Equals(beaconPrivateKey.PrivateKey.PublicKey()) {
+				return fmt.Errorf("configured beacon key is inconsistent with this node's canonical public beacon key (%s!=%s)",
+					beaconPrivateKey.PrivateKey.PublicKey().String(),
+					myBeaconPublicKeyShare.String())
```

Comment:
I guess this is equivalent
```suggestion
					beaconPrivateKey.PrivateKey.PublicKey(),
					myBeaconPublicKeyShare)
```